.github/workflows/inspektor-gadget.yml

```
@@ -240,6 +240,53 @@ jobs:
         cache-from: type=local,src=/tmp/.buildx-cache
         cache-to: type=local,dest=/tmp/.buildx-cache-new
 
+  build-gadget-core-container-image:
+    name: Build gadget CO-RE container image
+    # level: 0
+    runs-on: ubuntu-latest
+    steps:
+    - name: Set up Docker Buildx
+      id: buildx
+      uses: docker/setup-buildx-action@v1
+    - name: Cache Docker layers
+      uses: actions/cache@v2
+      with:
+        path: /tmp/.buildx-cache
+        key: ${{ runner.os }}-single-buildx-${{ hashFiles('gadget.core.Dockerfile') }}
+        restore-keys: |
+          ${{ runner.os }}-single-buildx
```

```suggestion
      uses: actions/cache@v2
      with:
        path: /tmp/.buildx-cache
        key: ${{ runner.os }}-buildx-gadget-${{ hashFiles('gadget.core.Dockerfile') }}
        restore-keys: |
          ${{ runner.os }}-buildx-gadget-
          ${{ runner.os }}-buildx-
```

We could use different cache prefix for the core image build and the non-core image build, since they are not based on the same image. Otherwise, in case of non-exact match on the key, we risk picking up the wrong one, depending on what is faster to build.

Similarly, the non-core image build could use `${{ runner.os }}-buildx-noncore-${{ hashFiles('gadget.Dockerfile') }}`

See https://docs.github.com/en/actions/using-workflows/caching-dependencies-to-speed-up-workflows#matching-a-cache-key

---

I'm no t sure I follow. What you're saying is that `restore-keys` should be different for both images to avoid picking up a wrong cache when there is not a match on `key`? 

Will using 

```        
restore-keys: |
          ${{ runner.os }}-core-single-buildx
```

and 

```        
restore-keys: |
          ${{ runner.os }}-bcc-single-buildx
```

fix the issue?

Also, do we really need these `restore-keys`?

---

> What you're saying is that restore-keys should be different for both images to avoid picking up a wrong cache when there is not a match on key?

Yes.

If a PR modifies the Dockerfile in some layer, the `${{ hashFiles(...) }}` will be different, but it is still worth falling back to an older cache from some older PR, but it has to be the same image, otherwise there will not be any Docker layer in common.

I think your suggested `restore-keys` should work.

> Also, do we really need these restore-keys?

As far as I understand, the `restore-keys` is just a fallback if the preferred `key` does not match. Having them will improve the perf if there is no exact match.

---

I checked the examples at https://github.com/actions/cache/blob/main/examples.md and it seems the restore keys are always used like the main key without `${{ hashFiles('...') }}`. I updated this to follow the same approach.